src/scattdata_header.F90

```diff
@@ -94,84 +116,169 @@ end subroutine scattdata_sample_
+      integer :: groups, gmin, gmax, gin
+      real(8) :: norm
 
       groups = size(energy, dim=1)
 
-      allocate(this % energy(groups, groups))
-      this % energy = energy
-      allocate(this % mult(groups, groups))
-      this % mult = mult
-      allocate(this % data(order, groups, groups))
-      this % data = ZERO
-
+      allocate(this % gmin(groups))
+      allocate(this % gmax(groups))
+      allocate(this % energy(groups))
+      allocate(this % mult(groups))
+      allocate(this % dist(groups))
+      ! Use energy to find the gmin and gmax values
+      ! Also set energy values when doing it
+      do gin = 1, groups
+        ! Make sure energy is normalized (i.e., CDF is 1)
+        norm = sum(energy(:,gin))
+        if (norm /= ZERO) energy(:,gin) = energy(:,gin) / norm
+        ! Find gmin by checking the P0 moment
+        do gmin = 1, groups
+          if (energy(gmin,gin) > ZERO) exit
+        end do
+        ! Find gmax by checking the P0 moment
+        do gmax = groups, 1, -1
+          if (energy(gmax,gin) > ZERO) exit
+        end do
+        ! Treat the case of all zeros
+        if (gmin > gmax) then
+          gmin = gin
+          gmax = gin
+          ! By not changing energy(gin) here we are leaving it as zero
+        end if
+        allocate(this % energy(gin) % data(gmin:gmax))
+        this % energy(gin) % data(gmin:gmax) = energy(gmin:gmax,gin)
+        allocate(this % mult(gin) % data(gmin:gmax))
+        this % mult(gin) % data(gmin:gmax) = mult(gmin:gmax,gin)
+        allocate(this % dist(gin) % data(order,gmin:gmax))
+        this % dist(gin) % data = ZERO
+        this % gmin(gin) = gmin
+        this % gmax(gin) = gmax
+      end do
     end subroutine scattdata_init
 
-    subroutine scattdatalegendre_init(this, order, energy, mult, coeffs)
-      class(ScattDataLegendre), intent(inout) :: this   ! Object to work on
-      integer, intent(in)                     :: order  ! Data Order
-      real(8), intent(in)              :: energy(:,:)   ! Energy Transfer Matrix
-      real(8), intent(in)              :: mult(:,:)     ! Scatter Prod'n Matrix
-      real(8), intent(in)              :: coeffs(:,:,:) ! Coefficients to use
-
-      real(8) :: dmu, mu, f
-      integer :: imu, Nmu, gout, gin, groups
+    subroutine scattdatalegendre_init(this, mult, coeffs)
+      class(ScattDataLegendre), intent(inout) :: this          ! Object to work on
+      real(8), intent(in)                     :: mult(:,:)     ! Scatter Prod'n Matrix
+      real(8), intent(in)                     :: coeffs(:,:,:) ! Coefficients to use
+
+      real(8) :: dmu, mu, f, norm
+      integer :: imu, Nmu, gout, gin, groups, order
+      real(8), allocatable :: energy(:,:)
+      real(8), allocatable :: matrix(:,:,:)
+
+      groups = size(coeffs,dim=3)
+      order = size(coeffs,dim=1)
+
+      ! make a copy of coeffs that we can use to extract data and normalize
+      allocate(matrix(order,groups,groups))
+      matrix = coeffs
```

Comment:
`matrix(:,:,:) = coeffs` is slightly safer. Technically F03+ is allowed to reallocate the LHS if you don't specify bounds like this.